utilities/vlogconf.8.in
.ds PN vlogconf

.TH vlogconf 8 "June 2008" "OpenFlow" "OpenFlow Manual"

.SH NAME
vlogconf \- configuration utility for OpenFlow logging in userspace

.SH SYNOPSIS
\fBvlogconf\fR [\fB-h\fR | \fB--help\fR] [\fItarget\fR...] [\fIaction\fR...]
.sp 1
The available \fItarget\fR options are:
.br
[\fB-a\fR | \fB--all\fR] [\fB-t\fR \fIpid\fR | \fB--target=\fIpid\fR]
.sp 1
The available \fIaction\fR options are:
.br
[\fB-l\fR | \fB--list\fR] [\fB-s\fR
\fImodule\fR[\fB:\fIfacility\fR[\fB:\fIlevel\fR]] |
\fB--set=\fImodule\fR[\fB:\fIfacility\fR[\fB:\fIlevel\fR]]]
[\fB-r\fR | \fB--reopen\fR]

.SH DESCRIPTION
The \fBvlogconf\fR program configures the logging system used by 
OpenFlow userspace programs.  The logging configuration may be modified 
while OpenFlow programs are running.

\fBvlogconf\fR applies one or more actions to each of one or more
target processes.  Targets may be specified as:

.TP
\fB-a\fR, \fB--all\fR
All running processes that \fBvlogconf\fR can control.

.TP
\fB-t \fItarget\fR, \fB--target=\fItarget\fR
The specified \fItarget\fR, which must take one of the following forms:

.RS
.IP \(bu
A PID (process ID).

.IP \(bu
An absolute path (beginning with `/') to the Unix domain socket for a
\fBvlogconf\fR-controllable process.

.IP \(bu
An absolute path (beginning with `/') to a pidfile (created by, e.g.,
passing the \fB-P\fR or \fB--pidfile\fR option to one of the OpenFlow
programs).  

.IP \(bu
None of the above, in which case \fItarget\fR prefixed by
\fB@RUNDIR@/\fR must match one of the cases for absolute paths listed
above.  (The default name for a program's pidfile is
\fB@RUNDIR@/\fIprogram\fB.pid\fR, so this means that, say,
\fBofprotocol\fR's default pidfile may be referred to simply as
\fBofprotocol.pid\fR.)
.RE

.PP
The available actions are:

.TP
\fB-l\fR, \fB--list\fR
Print the list of known modules and their current logging levels to
stdout.

.TP
\fB-s\fR \fImodule\fR[\fB:\fIfacility\fR[\fB:\fIlevel\fR]], \fB--set=\fImodule\fR[\fB:\fIfacility\fR[\fB:\fIlevel\fR]]

Sets the logging level for \fImodule\fR in \fIfacility\fR to
\fIlevel\fR.  The \fImodule\fR may be any valid module name (as
displayed by the \fB--list\fR option) or the special name \fBANY\fR to
set the logging levels for all modules.  The \fIfacility\fR may be
\fBsyslog\fR or \fBconsole\fR to set the levels for logging to the
system log or to the console, respectively, or \fBANY\fR to set the
logging levels for both facilities.  If it is omitted,
\fIfacility\fR defaults to \fBANY\fR.  The \fIlevel\fR must be one of
\fBemer\fR, \fBerr\fR, \fBwarn\fR, \fBinfo\fR, or \fBdbg\fR, designating the
minimum severity of a message for it to be logged.  If it is omitted,
\fIlevel\fR defaults to \fBdbg\fR.

.TP
\fB-s PATTERN:\fIfacility\fB:\fIpattern\fR, \fB--set=PATTERN:\fIfacility\fB:\fIpattern\fR

Sets the log pattern for \fIfacility\fR to \fIpattern\fR.  Each time a
message is logged to \fIfacility\fR, \fIpattern\fR determines the
message's formatting.  Most characters in \fIpattern\fR are copied
literally to the log, but special escapes beginning with \fB%\fR are
expanded as follows:

.RS
.TP
\fB%A\fR
The name of the application logging the message, e.g. \fBofprotocol\fR.

.TP
\fB%c\fR
The name of the module (as shown by \fBvlogconf --list\fR) logging
the message.

.TP
\fB%d\fR
The current date and time in ISO 8601 format (YYYY-MM-DD HH:MM:SS).

.TP
\fB%d{\fIformat\fB}\fR
The current date and time in the specified \fIformat\fR, which takes
the same format as the \fItemplate\fR argument to \fBstrftime\fR(3).

.TP
\fB%m\fR
The message being logged.

.TP
\fB%N\fR
A serial number for this message within this run of the program, as a
decimal number.  The first message a program logs has serial number 1,
the second one has serial number 2, and so on.

.TP
\fB%n\fR
A new-line.

.TP
\fB%p\fR
The level at which the message is logged, e.g. \fBDBG\fR.

.TP
\fB%P\fR
The program's process ID (pid), as a decimal number.

.TP
\fB%r\fR
The number of milliseconds elapsed from the start of the application
to the time the message was logged.

.TP
\fB%%\fR
A literal \fB%\fR.
.RE

.IP
A few options may appear between the \fB%\fR and the format specifier
character, in this order:

.RS
.TP
\fB-\fR
Left justify the escape's expansion within its field width.  Right
justification is the default.

.TP
\fB0\fR
Pad the field to the field width with \fB0\fRs.  Padding with spaces
is the default.

.TP
\fIwidth\fR
A number specifies the minimum field width.  If the escape expands to
fewer characters than \fIwidth\fR then it is padded to fill the field
width.  (A field wider than \fIwidth\fR is not truncated to fit.)
.RE

.IP
The default pattern for console output is \fB%d{%b %d
%H:%M:%S}|%05N|%c|%p|%m\fR; for syslog output, \fB%05N|%c|%p|%m\fR.

.TP
\fB-r\fR, \fB--reopen\fR
Causes the target application to close and reopen its log file.  (This
is useful after rotating log files, to cause a new log file to be
used.)

.SH OPTIONS

.so lib/common.man

.SH "SEE ALSO"

.BR dpctl (8),
.BR ofprotocol (8),
.BR controller (8)


debian/changelog
openflow (1.0.0) unstable; urgency=low

  * Development version

 -- OpenFlow team <openflow-dev@lists.stanford.edu>  Thu, 31 Dec 2009 23:59:59 -0800

openflow (0.9.0-rev1) unstable; urgency=low

  * Development version.

 -- OpenFlow team <openflow-dev@lists.stanford.edu>  Fri, 04 Sep 2009 12:00:00 -0800

openflow (0.8.9-rev4) unstable; urgency=low

  * Develoment version.

 -- OpenFlow team <openflow-dev@lists.stanford.edu>  Tue, 15 Sep 2009 12:00:00 -0800

openflow (0.8.1) unstable; urgency=low

  * Development version.

 -- OpenFlow team <openflow-dev@lists.stanford.edu>  Mon, 19 Nov 2007 14:57:52 -0800


utilities/vlogconf.c
/* Copyright (c) 2008, 2009 The Board of Trustees of The Leland Stanford
 * Junior University
 * 
 * We are making the OpenFlow specification and associated documentation
 * (Software) available for public use and benefit with the expectation
 * that others will use, modify and enhance the Software and contribute
 * those enhancements back to the community. However, since we would
 * like to make the Software available for broadest use, with as few
 * restrictions as possible permission is hereby granted, free of
 * charge, to any person obtaining a copy of this Software to deal in
 * the Software under the copyrights without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * The name and trademarks of copyright holder(s) may NOT be used in
 * advertising or publicity pertaining to the Software or any
 * derivatives without specific, written prior permission.
 */
#include <config.h>
#include "vlog.h"

#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "command-line.h"
#include "compiler.h"
#include "timeval.h"
#include "util.h"
#include "vlog-socket.h"

static void
usage(char *prog_name, int exit_code)
{
    printf("Usage: %s [TARGET] [ACTION...]\n"
           "Targets:\n"
           "  -a, --all            Apply to all targets (default)\n"
           "  -t, --target=TARGET  Specify target program, as a pid, a\n"
           "                       pidfile, or an absolute path to a Unix\n"
           "                       domain socket\n"
           "Actions:\n"
           "  -l, --list         List current settings\n"
           "  -s, --set=MODULE[:FACILITY[:LEVEL]]\n"
           "        Set MODULE and FACILITY log level to LEVEL\n"
           "        MODULE may be any valid module name or 'ANY'\n"
           "        FACILITY may be 'syslog', 'console', 'file', or 'ANY' (default)\n"
           "        LEVEL may be 'emer', 'err', 'warn', 'info', or 'dbg' (default)\n"
           "  -r, --reopen       Make the program reopen its log file\n"
           "  -h, --help         Print this helpful information\n",
           prog_name);
    exit(exit_code);
}

static char *
transact(struct vlog_client *client, const char *request, bool *ok)
{
    char *reply;
    int error = vlog_client_transact(client, request, &reply);
    if (error) {
        fprintf(stderr, "%s: transaction error: %s\n",
                vlog_client_target(client), strerror(error));
        *ok = false;
    }
    return reply ? reply : xstrdup("");
}

static void
transact_ack(struct vlog_client *client, const char* request, bool *ok)
{
    char *reply;
    int error = vlog_client_transact(client, request, &reply);
    if (error) {
        fprintf(stderr, "%s: transaction error: %s\n",
                vlog_client_target(client), strerror(error));
        *ok = false;
    } else if (strcmp(reply, "ack")) {
        fprintf(stderr, "Received unexpected reply from %s: %s\n",
                vlog_client_target(client), reply);
        *ok = false;
    }
    free(reply);
}

static void
add_target(struct vlog_client ***clients, size_t *n_clients,
           const char *path, bool *ok)
{
    struct vlog_client *client;
    int error = vlog_client_connect(path, &client);
    if (error) {
        fprintf(stderr, "Error connecting to \"%s\": %s\n",
                path, strerror(error));
        *ok = false;
    } else {
        *clients = xrealloc(*clients, sizeof *clients * (*n_clients + 1));
        (*clients)[*n_clients] = client;
        ++*n_clients;
    }
}

static void
add_all_targets(struct vlog_client ***clients, size_t *n_clients, bool *ok)
{
    DIR *directory;
    struct dirent* de;

    directory = opendir("/tmp");
    if (!directory) {
        fprintf(stderr, "/tmp: opendir: %s\n", strerror(errno));
    }

    while ((de = readdir(directory)) != NULL) {
        if (!strncmp(de->d_name, "vlogs.", 5)) {
            char *path = xasprintf("/tmp/%s", de->d_name);
            add_target(clients, n_clients, path, ok);
            free(path);
        }
    }

    closedir(directory);
}

int main(int argc, char *argv[])
{
    static const struct option long_options[] = {
        /* Target options must come first. */
        {"all", no_argument, NULL, 'a'},
        {"target", required_argument, NULL, 't'},
        {"help", no_argument, NULL, 'h'},

        /* Action options come afterward. */
        {"list", no_argument, NULL, 'l'},
        {"set", required_argument, NULL, 's'},
        {"reopen", no_argument, NULL, 'r'},
        {0, 0, 0, 0},
    };
    char *short_options;

    /* Determine targets. */
    bool ok = true;
    int n_actions = 0;
    struct vlog_client **clients = NULL;
    size_t n_clients = 0;

    set_program_name(argv[0]);
    time_init();

    short_options = long_options_to_short_options(long_options);
    for (;;) {
        int option;
        size_t i;

        option = getopt_long(argc, argv, short_options, long_options, NULL);
        if (option == -1) {
            break;
        }
        if (!strchr("ath", option) && n_clients == 0) {
            ofp_fatal(0, "no targets specified (use --help for help)");
        } else {
            ++n_actions;
        }
        switch (option) {
        case 'a':
            add_all_targets(&clients, &n_clients, &ok);
            break;

        case 't':
            add_target(&clients, &n_clients, optarg, &ok);
            break;

        case 'l':
            for (i = 0; i < n_clients; i++) {
                struct vlog_client *client = clients[i];
                char *reply;

                printf("%s:\n", vlog_client_target(client));
                reply = transact(client, "list", &ok);
                fputs(reply, stdout);
                free(reply);
            }
            break;

        case 's':
            for (i = 0; i < n_clients; i++) {
                struct vlog_client *client = clients[i];
                char *request = xasprintf("set %s", optarg);
                transact_ack(client, request, &ok);
                free(request);
            }
            break;

        case 'r':
            for (i = 0; i < n_clients; i++) {
                struct vlog_client *client = clients[i];
                char *request = xstrdup("reopen");
                transact_ack(client, request, &ok);
                free(request);
            }
            break;

        case 'h':
            usage(argv[0], EXIT_SUCCESS);
            break;

        case '?':
            exit(EXIT_FAILURE);

        default:
            NOT_REACHED();
        }
    }
    if (!n_actions) {
        fprintf(stderr,
                "warning: no actions specified (use --help for help)\n");
    }
    exit(ok ? 0 : 1);
}
